openai_api.py

```diff
@@ -0,0 +1,596 @@
+                            yield f"data: {response_json}\n\n"
+                            delta_q = []  # Clear delta_q for the next iteration
+                            token_ids = None  # Clear inputs for the next generation step
+
+                        if len(delta) > 0 and delta[-1] == tokenizer.eos_token_id:
+                            yield END_STRING
+                            return
+
+                yield END_STRING
+
+        return StreamingResponse(completion_stream_generator(model, token_ids),
+                                 media_type="text/event-stream")
+
+    num_prompt_tokens = token_ids.shape[1]
+
+    results_generator = model.generate(token_ids, **generation_config.to_diff_dict())
+
+    choices = []
+    for index, output_tokens in enumerate(results_generator):
+        text = tokenizer.decode(
+            output_tokens[num_prompt_tokens:],
+            skip_special_tokens=skip_special_tokens,
+            spaces_between_special_tokens=spaces_between_special_tokens,
+        )
+
+        choice_data = ChatCompletionResponseChoice(
+            index=index,
+            message=ChatMessage(role="assistant", content=text),
+            finish_reason=None,
+        )
+        choices.append(choice_data)
+
+    num_generated_tokens = sum(
+        len(output) for output in results_generator)
+    usage = UsageInfo(
+        prompt_tokens=num_prompt_tokens,
+        completion_tokens=num_generated_tokens,
+        total_tokens=num_prompt_tokens + num_generated_tokens,
+    )
+    response = ChatCompletionResponse(
+        id=request_id,
+        created=created_time,
+        model=model_name,
+        choices=choices,
+        usage=usage,
+    )
+
+    if request.stream:
+        # When user requests streaming but we don't stream, we still need to
+        # return a streaming response with a single event.
+        response_json = response.json(ensure_ascii=False)
+
+        async def fake_stream_generator() -> AsyncGenerator[str, None]:
+            yield f"data: {response_json}\n\n"
+            yield "data: [DONE]\n\n"
+
+        return StreamingResponse(fake_stream_generator(),
+                                 media_type="text/event-stream")
+
+    return response
+
+
+@app.post("/v1/completions")
+async def create_completion(request: CompletionRequest, raw_request: Request):
+    """Completion API similar to OpenAI's API.
+
+    See https://platform.openai.com/docs/api-reference/completions/create
+    for the API specification. This API mimics the OpenAI Completion API.
+
+    NOTE: Currently we do not support the following features:
+        - echo (since the Petals engine does not currently support
+          getting the logprobs of prompt tokens)
+        - suffix (the language models we currently support do not support
+          suffix)
+        - logit_bias (to be supported by Petals engine)
+        - stream (to be supported by Petals engine)
```

Comment:
Is stream is not supported? 🤨

Reply:
Docstring and code are not aligned with each other.

Reply:
Updated docstring info